apps/emqx/src/emqx_router_helper.erl

```diff
@@ -192,5 +122,61 @@ code_change(_OldVsn, State, _Extra) ->
 stats_fun() ->
     emqx_stats:setstat('topics.count', 'topics.max', emqx_router:stats(n_routes)).
 
+%% This is just to avoid cleaning up routes twice in case the node is down,
+%% as we can expect {mria, down}, {node, down} under normal circumstances
+%%  when a replicant node is being stopped.
+maybe_drain_nodedown_msg(What, DownNode) when DownNode =/= node(), What =/= node ->
+    receive
+        {membership, {node, down, DownNode}} -> ok
+    after ?DRAIN_TIMEOUT -> ok
+    end;
+maybe_drain_nodedown_msg(_, _) ->
+    ok.
+
+locking_nodes() ->
+    locking_nodes(mria_rlog:role()).
```

Comment:
this and `mria_config:whoami()`, which is the correct API ?

Reply:
I think both are equally usable. `whoami/0` will tel if the rlog is disabled and will return `mnesia`. 
`role/0` always return `core` or `replicant`.